lib/typing_env.ml

```diff
@@ -1,50 +1,121 @@
+        | Sig_value (id, vd, v) as sig_val -> (
+            match Env.find_value_index id ref_env with
+            | Some _ ->
+                let new_id = Ident.rename id in
+                (subst, Sig_value (new_id, vd, v) :: lst)
+            | None -> (subst, sig_val :: lst))
+        | Sig_class (id, cd, r, v) as sig_cls_decl -> (
+            match Env.find_class_index id ref_env with
+            | Some _ ->
+                let new_id = Ident.rename id in
+                (subst, Sig_class (new_id, cd, r, v) :: lst)
+            | None -> (subst, sig_cls_decl :: lst))
+        | Sig_class_type (id, ct, r, v) as sig_cltype -> (
+            match Env.find_cltype_index id ref_env with
+            | Some _ ->
+                let new_id = Ident.rename id in
+                (subst, Sig_class_type (new_id, ct, r, v) :: lst)
+            | None -> (subst, sig_cltype :: lst))
+        | _ -> (subst, item :: lst))
+      current (Subst.identity, [])
+  in
+  List.map
+    (function
+      | Sig_type (id, td, r, v) ->
+          Sig_type (id, Subst.type_declaration subst td, r, v)
+      | Sig_value (id, vd, v) ->
+          Sig_value (id, Subst.value_description subst vd, v)
+      | Sig_class (id, cd, rc, v) ->
+          Sig_class (id, Subst.class_declaration subst cd, rc, v)
+      | Sig_class_type (id, ct, rc, v) ->
+          Sig_class_type (id, Subst.cltype_declaration subst ct, rc, v)
+      | Sig_modtype (id, m, v) ->
+          Sig_modtype (id, Subst.modtype_declaration Subst.Keep subst m, v)
+      | Sig_module (id, mp, md, rc, v) ->
+          Sig_module
+            (id, mp, Subst.module_declaration Subst.Keep subst md, rc, v)
+      | Sig_typext (id, ec, es, v) ->
+          Sig_typext (id, Subst.extension_constructor subst ec, es, v))
+    modified_current
+
+let extract_subst_items signature =
   List.fold_left
     (fun acc item ->
       match item with
-      | Sig_type (id, ({ type_manifest = None; _ } as type_decl), _, _) ->
-          String_map.add (Ident.name id) (id, type_decl) acc
+      | Sig_type (id, { type_manifest = None; _ }, _, _) ->
+          Subst_item_map.add (Type, Ident.name id) id acc
+      | Sig_module (id, _, _, _, _) ->
+          Subst_item_map.add (Module, Ident.name id) id acc
+      | Sig_modtype (id, _, _) ->
+          Subst_item_map.add (Modtype, Ident.name id) id acc
       | _ -> acc)
-    String_map.empty signature
-
-let set_type_equality ref_id ~ref_decl ~curr_decl =
-  if List.length ref_decl.type_params = List.length curr_decl.type_params then
-    let ref_path = Path.Pident ref_id in
-    {
-      curr_decl with
-      type_manifest =
-        Some
-          (Btype.newgenty (Tconstr (ref_path, curr_decl.type_params, ref Mnil)));
-    }
-  else curr_decl
+    Subst_item_map.empty signature
 
-let set_type_equalities ~reference ~current =
-  let ref_non_alias_types = extract_non_alias_types reference in
-  List.map
-    (fun item ->
+let items_subst ~reference ~current =
```

Comment:
This function is creating a `Subst.t` that's meant to be used to replace types and other such items that should be considered equal is that correct?

I think this could also use a more meaningful name and a tiny bit of documentation!

Reply:
Yes. It think it was called ```pair_components``` in the compiler libs, Is that a good name ?

Reply:
Yeah that sounds good!